src/idlcxx/src/traits.c

```diff
@@ -72,6 +72,21 @@ write_blob(
 }
 
 
+static bool
+emit_isKeyless(
+  const idl_pstate_t* pstate,
+  const void* node)
+{
+  if (!idl_is_struct(node)) {
+    return idl_is_keyless(node, pstate->config.flags & IDL_FLAG_KEYLIST);
+  } else {
+    const idl_struct_t * _struct = node;
+    while (_struct->inherit_spec)
+      _struct = _struct->inherit_spec->base;
+    return idl_is_keyless(_struct, pstate->config.flags & IDL_FLAG_KEYLIST);
+  }
```

Comment:
Thanks @reicheratwork. In "7.2.2.4.5 Inheritance of Aggregated Types" the spec states (among other things):

> The derived type does not define any key fields. This ensures the key fields of the derived type are the same as those of the base root type.

So it shouldn't make any difference with the current version of XTypes. I don't think this detail is likely to change anytime soon but I am not convinced it will never change. Might as well use your version now 🙂

Reply:
On second thought (after changing the code and thinking about a commit message), I think it is better to leave it as is: *if* the spec ever changes, that'll involve a lot more work than this one function and it is reasonable to assume that the relevant changes would be done as part of making it compatible with this hypothetical new version of the spec.

What seems more likely is that the `idl_is_keyless` function will get modified to take inheritance into account. In that case, this entire function becomes superfluous and forgetting to update it (an easy mistake to make and easily overlooked in a review) might well result in an O(n²) check.

Perhaps I should do that change to `idl_is_keyless` now, but I'd like to first see what the consequences of that might be that and I do want this fixed.